pubsub.go

```diff
@@ -668,6 +727,48 @@ func (p *PubSub) publishMessage(from peer.ID, pmsg *pb.Message) {
 	p.rt.Publish(from, pmsg)
 }
 
+type addTopicReq struct {
+	topic *Topic
+	resp  chan *Topic
+}
+
+type rmTopicReq struct {
+	topic *Topic
+	resp  chan error
+}
+
+type TopicOptions struct{}
+
+type TopicOpt func(t *Topic) error
+
+func (p *PubSub) Join(topic string, opts ...TopicOpt) (*Topic, error) {
+	t := &Topic{
+		p:           p,
+		topic:       topic,
+		evtHandlers: make(map[*TopicEventHandler]struct{}),
+	}
+
+	for _, opt := range opts {
+		err := opt(t)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	resp := make(chan *Topic, 1)
+	t.p.addTopic <- &addTopicReq{
+		topic: t,
+		resp:  resp,
+	}
+	returnedTopic := <-resp
+
+	if returnedTopic != t {
+		return returnedTopic, fmt.Errorf("topic already exists")
```

Comment:
Sure, we're assuming a single/exclusive topic handle here (as described in #198). I'm happy to remove the error (it felt a bit wrong anyway), but do you think it's useful for a service to know that when they `Join` a topic that they weren't the first ones?

Seems like it could potentially come in handy, but certainly isn't necessary.

Reply:
(documenting the results of a sync call)

I was suggesting that we remove the _topic_, not the error. The alternative is to make topics multi-subscriber friendly but that would involve a bunch of work to handle conflicting validators, reference counts, etc.

Reply:
That's fine we can return nil. The tradeoff will be that we'll need to expose a private function that allows us to get another reference to the topic so that we can continue to use the existing `Publish` and `Subscribe` functions.

Sound good? Also, should I put a deprecation warning on `Publish` and `Subscribe`?

Reply:
SGTM and yes, we should deprecate (IMO).